packages/core/src/browser/source-tree/source-tree.ts

```diff
@@ -41,50 +43,56 @@ export class SourceTree extends TreeImpl {
         return parent.element.getElements();
     }
 
-    protected toNode(element: TreeElement, index: number, parent: TreeElementNodeParent): TreeElementNode {
-        const id = element.id ? String(element.id) : (parent.id + ':' + index);
-        const name = id;
-        const existing = this.getNode(id);
-        const updated = existing && <TreeElementNode>Object.assign(existing, { element, parent });
+    /**
+     * Convert a parent's elements into `TreeElementNode` or `CompositeTreeElementNode`.
+     *
+     * If a node already exists in this tree, we'll mutate it rather than create a new updated one.
+     */
+    protected toNode(element: TreeElement, index: number, parent: TreeElementNodeParent): TreeElementNode | CompositeTreeElementNode {
+        const id: string = element.id === undefined ? `${parent.id}:${index}` : String(element.id);
+        // The base `TreeImpl` class expects to store simple `TreeNode` instances,
+        // but we know for a fact that we handle `TreeElementNode` in this `SourceTree` class.
+        // TODO: `Tree` should be generic and require something like `T extends TreeNode`?
+        const existing = this.getNode(id) as TreeElementNode | CompositeTreeElementNode | undefined;
+        if (existing) {
+            existing.element = element;
+            existing.parent = parent;
+        }
         if (CompositeTreeElement.hasElements(element)) {
-            if (updated) {
-                if (!ExpandableTreeNode.is(updated)) {
-                    Object.assign(updated, { expanded: false });
+            if (existing) {
+                if (!ExpandableTreeNode.is(existing)) {
+                    (existing as any).expanded = false;
                 }
-                if (!CompositeTreeNode.is(updated)) {
-                    Object.assign(updated, { children: [] });
+                if (!CompositeTreeNode.is(existing)) {
+                    (existing as any).children = [];
                 }
-                return updated;
+                return existing;
             }
             return {
                 element,
                 parent,
                 id,
-                name,
+                name: id,
                 selected: false,
                 expanded: false,
                 children: []
-            } as TreeElementNode;
-        }
-        if (CompositeTreeElementNode.is(updated)) {
-            delete updated.expanded;
-            delete updated.children;
+            };
         }
-        if (updated) {
-            if (ExpandableTreeNode.is(updated)) {
-                delete updated.expanded;
+        if (existing) {
+            if (ExpandableTreeNode.is(existing)) {
+                delete (existing as any).expanded;
```

Comment:
Rather than an `any` cast here, you can use `Partial<ExpandableTreeNode>` (or `CompositeTreeNode`, where appropriate). I'd prefer to avoid file-scoped disabling of `no-explicit-any` where possible.